### DISCUSSION TITLE
Import carla SystemError

### DISCUSSION THREAD

**Discussion by JoFeGoOr** [carla-simulator/carla#6809] [Q&A] [answered]
Hello, i am having some problems with carla 0.9.14 on windows 11

i am following the Windows build guide, i was able to launch the editor, but i got some problems starting the simulation.

1.- when i try to run an example from PythonAPI\examples folder, i get an error when i try to install the requirements.txt i get this:

------------------------------------------
C:\carla\PythonAPI\examples>pip install -r requirements.txt
Defaulting to user installation because normal site-packages is not writeable
Ignoring numpy: markers 'python_version < "3.0"' don't match your environment
Requirement already satisfied: future in c:\users\danielyunge\appdata\roaming\python\python312\site-packages (from -r requirements.txt (line 1)) (0.18.3)
WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000001ECCD6A5CA0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/numpy/
WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000001ECCD6A6900>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/numpy/
Collecting numpy==1.18.4 (from -r requirements.txt (line 3))
  Using cached numpy-1.18.4.zip (5.4 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [68 lines of output]
      Running from numpy source directory.
      <string>:461: UserWarning: Unrecognized setuptools command, proceeding with generating Cython sources and expanding templates
      C:\Users\DanielYunge\AppData\Local\Temp\pip-install-_2idxxte\numpy_2fe58d9fa39e442a94853eb88d03c76e\tools\cythonize.py:75: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
        required_version = LooseVersion('0.29.14')
      C:\Users\DanielYunge\AppData\Local\Temp\pip-install-_2idxxte\numpy_2fe58d9fa39e442a94853eb88d03c76e\tools\cythonize.py:77: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
        if LooseVersion(cython_version) < required_version:
      warning: C:\Users\DanielYunge\AppData\Local\Temp\pip-install-_2idxxte\numpy_2fe58d9fa39e442a94853eb88d03c76e\numpy\__init__.pxd:17:0: The 'DEF' statement is deprecated and will be removed in a future Cython version. Consider using global variables, constants, and in-place literals instead. See https://github.com/cython/cython/issues/4310
      warning: C:\Users\DanielYunge\AppData\Local\Temp\pip-install-_2idxxte\numpy_2fe58d9fa39e442a94853eb88d03c76e\numpy\__init__.pxd:17:0: The 'DEF' statement is deprecated and will be removed in a future Cython version. Consider using global variables, constants, and in-place literals instead. See https://github.com/cython/cython/issues/4310
      warning: C:\Users\DanielYunge\AppData\Local\Temp\pip-install-_2idxxte\numpy_2fe58d9fa39e442a94853eb88d03c76e\numpy\__init__.pxd:17:0: The 'DEF' statement is deprecated and will be removed in a future Cython version. Consider using global variables, constants, and in-place literals instead. See https://github.com/cython/cython/issues/4310
      warning: C:\Users\DanielYunge\AppData\Local\Temp\pip-install-_2idxxte\numpy_2fe58d9fa39e442a94853eb88d03c76e\numpy\__init__.pxd:17:0: The 'DEF' statement is deprecated and will be removed in a future Cython version. Consider using global variables, constants, and in-place literals instead. See https://github.com/cython/cython/issues/4310
      warning: _generator.pyx:3330:8: The 'DEF' statement is deprecated and will be removed in a future Cython version. Consider using global variables, constants, and in-place literals instead. See https://github.com/cython/cython/issues/4310
      warning: C:\Users\DanielYunge\AppData\Local\Temp\pip-install-_2idxxte\numpy_2fe58d9fa39e442a94853eb88d03c76e\numpy\__init__.pxd:17:0: The 'DEF' statement is deprecated and will be removed in a future Cython version. Consider using global variables, constants, and in-place literals instead. See https://github.com/cython/cython/issues/4310
      warning: C:\Users\DanielYunge\AppData\Local\Temp\pip-install-_2idxxte\numpy_2fe58d9fa39e442a94853eb88d03c76e\numpy\__init__.pxd:17:0: The 'DEF' statement is deprecated and will be removed in a future Cython version. Consider using global variables, constants, and in-place literals instead. See https://github.com/cython/cython/issues/4310

      Error compiling Cython file:
      ------------------------------------------------------------
      ...
              for i in range(1, RK_STATE_LEN):
                  self.rng_state.key[i] = val[i]
              self.rng_state.pos = i

              self._bitgen.state = &self.rng_state
              self._bitgen.next_uint64 = &mt19937_uint64
                                         ^
      ------------------------------------------------------------

      _mt19937.pyx:138:35: Cannot assign type 'uint64_t (*)(void *) except? -1 nogil' to 'uint64_t (*)(void *) noexcept nogil'. Exception values are incompatible. Suggest adding 'noexcept' to type 'uint64_t (void *) except? -1 nogil'.
      Processing numpy/random\_bounded_integers.pxd.in
      Processing numpy/random\mtrand.pyx
      Processing numpy/random\_bit_generator.pyx
      Processing numpy/random\_bounded_integers.pyx.in
      Processing numpy/random\_common.pyx
      Processing numpy/random\_generator.pyx
      Processing numpy/random\_mt19937.pyx
      Traceback (most recent call last):
        File "C:\Users\DanielYunge\AppData\Local\Temp\pip-install-_2idxxte\numpy_2fe58d9fa39e442a94853eb88d03c76e\tools\cythonize.py", line 238, in <module>
          main()
        File "C:\Users\DanielYunge\AppData\Local\Temp\pip-install-_2idxxte\numpy_2fe58d9fa39e442a94853eb88d03c76e\tools\cythonize.py", line 234, in main
          find_process_files(root_dir)
        File "C:\Users\DanielYunge\AppData\Local\Temp\pip-install-_2idxxte\numpy_2fe58d9fa39e442a94853eb88d03c76e\tools\cythonize.py", line 225, in find_process_files
          process(root_dir, fromfile, tofile, function, hash_db)
        File "C:\Users\DanielYunge\AppData\Local\Temp\pip-install-_2idxxte\numpy_2fe58d9fa39e442a94853eb88d03c76e\tools\cythonize.py", line 191, in process
          processor_function(fromfile, tofile)
        File "C:\Users\DanielYunge\AppData\Local\Temp\pip-install-_2idxxte\numpy_2fe58d9fa39e442a94853eb88d03c76e\tools\cythonize.py", line 80, in process_pyx
          subprocess.check_call(
        File "C:\Program Files\Python312\Lib\subprocess.py", line 413, in check_call
          raise CalledProcessError(retcode, cmd)
      subprocess.CalledProcessError: Command '['C:\\Program Files\\Python312\\python.exe', '-m', 'cython', '-3', '--fast-fail', '-o', '_mt19937.c', '_mt19937.pyx']' returned non-zero exit status 1.
      Cythonizing sources
      Traceback (most recent call last):
        File "C:\Program Files\Python312\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 353, in <module>
          main()
        File "C:\Program Files\Python312\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Program Files\Python312\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 149, in prepare_metadata_for_build_wheel
          return hook(metadata_directory, config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\DanielYunge\AppData\Local\Temp\pip-build-env-2y1onhit\overlay\Lib\site-packages\setuptools\build_meta.py", line 396, in prepare_metadata_for_build_wheel
          self.run_setup()
        File "C:\Users\DanielYunge\AppData\Local\Temp\pip-build-env-2y1onhit\overlay\Lib\site-packages\setuptools\build_meta.py", line 507, in run_setup
          super(_BuildMetaLegacyBackend, self).run_setup(setup_script=setup_script)
        File "C:\Users\DanielYunge\AppData\Local\Temp\pip-build-env-2y1onhit\overlay\Lib\site-packages\setuptools\build_meta.py", line 341, in run_setup
          exec(code, locals())
        File "<string>", line 488, in <module>
        File "<string>", line 469, in setup_package
        File "<string>", line 275, in generate_cython
      RuntimeError: Running cythonize failed!
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

--------------------------------------------------------------------

2.- i tried to run some examples anyway and i got this error:

-------------------------------
C:\carla\PythonAPI\examples>python -m generate_traffic.py
Traceback (most recent call last):
  File "<frozen runpy>", line 189, in _run_module_as_main
  File "<frozen runpy>", line 112, in _get_module_details
  File "C:\carla\PythonAPI\examples\generate_traffic.py", line 24, in <module>
    import carla
  File "C:\Users\DanielYunge\AppData\Roaming\Python\Python312\site-packages\carla\__init__.py", line 8, in <module>
    from .libcarla import *
SystemError: type Boost.Python.enum has the Py_TPFLAGS_HAVE_GC flag but has no traverse function

------------------------------------------------------------------------------

I also got this error when i try to run import carla in jupyterlabs

3.- from the requirements.txt i install everything manually and got an error when i run pip install open 3d:

------------------------------------------------
C:\carla\PythonAPI\examples>pip install open3d
Defaulting to user installation because normal site-packages is not writeable
ERROR: Could not find a version that satisfies the requirement open3d (from versions: none)
ERROR: No matching distribution found for open3d

-----------------------------------------------------

sorry if this is a lot, i have not been able to find solutions online

thanks for your time.

**Comment by alahi45** [comment]
Same error.  I have also tried with python3 (as mentioned in the guide), python, py. But nothing seemed to work. Maybe this has something to do with the .egg or .whl file. I'm not sure. Let me know if you find any solution.

**Comment by JoFeGoOr** [comment, marked as answer]
i repeated the whole installation but i install python 3.9, also i make sure that the "dist" folder was created and finally from the "requirements.txt" i did the installations myself.

i was able to execute "python manual_control.py"

**Reply by alahi45** [reply]
Hi, Thanks for your response. During installation, "dist" folder was created and there are two files with .whl and .egg. But still I'm getting the same error. 
Have you run this "pip3 install <path/to/wheel>.whl" after "make PythonAPI" command? 
Did you install the exact version mentioned in the "requirements.txt"?

**Reply by CMakey** [reply]
after see this, instead of repeating the whole installation, I only downgrade the python, but now the it can't generate the `dist`folder , and even I download the wheel from other places, i still can't install numpy==1.18.4

**Comment by JoFeGoOr** [comment]
> Have you run this "pip3 install <path/to/wheel>.whl" after "make PythonAPI" command?

yes, in the dist folder i run "pip install <wheel_file_name>.whl" after the "make PythonAPI", remember to use the x64 Command Prompt

>Did you install the exact version mentioned in the "requirements.txt"?

i try but some of them did not work with the exact version mentioned in the "requirements.txt" so i did some googleling to fix those

it's worth mention that running "python generate_traffic.py" freezes the editor and i dont know if it is related with the steps i took earlier

**Reply by alahi45** [reply]
Thanks. I have downgraded to python 3.8 and repeated the whole procedure. 
Now everything is working fine. I was also able to run "python generate_traffic.py". What are you working on? Maybe we can connect to each other and share our insights.

**Reply by JoFeGoOr** [reply]
Hi, im working on roadrunner right now, i am having some problems importing a map to the carla simulator, i have try the "make import" method but does not seems to work, i am look some alternatives but nothing solid.

**Reply by alahi45** [reply]
Great! We are in the same boat then. I was also thinking of importing my own customized map. Unfortunately, I don't have the licensed version of Roadrunner, so I will try to use Truevision Designer ( Alternative of Roadrunner ).

**Reply by sgbtan** [reply]
Hi there, I'm having the same issue. 

I get the same error when installing requirements.txt 

Doing -m generate_traffic.py says "No module named 'carla'"

Any tips?

**Reply by evoevo07** [reply]
@alahi45 
Do you have any update, working with Truevision?
I am also searching for a good alternative.
